EVENTS.md

```diff
@@ -0,0 +1,416 @@
+at some good examples of what they look like:
+
+The most basic one is `bank`, which emits two events on every send, a [custom "transfer" event](https://github.com/cosmos/cosmos-sdk/blob/v0.42.9/x/bank/keeper/send.go#L142-L147) as well as "sender" information under the [standard "message" type](https://github.com/cosmos/cosmos-sdk/blob/v0.42.9/x/bank/keeper/send.go#L148-L151). Replacing variables with string literals, they look like this:
+
+```go
+sdk.NewEvent(
+    "transfer"
+    sdk.NewAttribute("recipient", toAddr.String()),
+    sdk.NewAttribute("sender", fromAddr.String()),
+    sdk.NewAttribute("amount", amt.String()),  // eg 12456uatom
+),
+sdk.NewEvent(
+    "message",
+    sdk.NewAttribute("sender", fromAddr.String()),
+),
+```
+
+The delegation module seems a bit more refined, emitting a generic "message" type event in [`msg_server.go`](https://github.com/cosmos/cosmos-sdk/blob/v0.42.9/x/distribution/keeper/msg_server.go#L42-L46) including the module name, **before** 
+emitting some custom event types closer to the actual code logic in
+[`keeper.go`](https://github.com/cosmos/cosmos-sdk/blob/v0.42.9/x/distribution/keeper/keeper.go#L74-L77).
+
+This looks something like:
+
+```go
+sdk.NewEvent(
+    "message",
+    sdk.NewAttribute("module", "distribution"),
+    sdk.NewAttribute("sender", msg.DelegatorAddress),
+),
+sdk.NewEvent(
+    "set_withdraw_address",
+    sdk.NewAttribute("withdraw_address", withdrawAddr.String()),
+),
+```
+
+## Usage in wasmd
+
+In `x/wasm` we also use Events system. On one hand, the Go implementation of `x/wasm` emits standard events for each 
+message it processes, using the `distribution` module as an example. Furthermore, it allows contracts to
+emit custom events based on their execution state, so they can for example say "dex swap, BTC-ATOM, in 0.23, out 512"
+which require internal knowledge of the contract and is very useful for custom dApp UIs.
+
+`x/wasm` is also a consumer of events, since when a smart contract executes a SubMsg and processes the reply, it receives
+not only the `data` response from the message execution, but also the list of events. This makes it even more important for
+us to document a standard event processing format.
+
+### Standard Events in x/wasm
+
+Following the model of `distribution`, we will split the emitted events into two parts. All calls to the message server, will receive
+the following event:
+
+```go
+sdk.NewEvent(
+    "message",
+    sdk.NewAttribute("module", "wasm"),
+    // Note: this was "signer" before 0.18
+    sdk.NewAttribute("sender", msg.Sender),  
+),
+```
+
+No further information will be added to the generic "message" type, but rather be contained in a more context-specific event type.
+Here are some examples:
+
+```go
+// Store Code
+sdk.NewEvent(
+    "store_code",
+    // Update in 0.18: _code_id is also a reserved prefix
```

Comment:
Do we need to make it a reserved prefix". For custom events we are injecting the contract address so there was the requirement but we are not doing the same for code_id. Maybe flattening?

Reply:
Let's leave this as code_id, as we will not be injecting it into `wasm*` events